### ISSUE TITLE
[Compiler][luci] Upgrade logic of shape/type inference algorithm

### ISSUE THREAD

**Issue by llFreetimell** [Samsung/ONE#5683]
## What?

Let's re-implement(upgrade) shape and type inference algorithm for following things!

- Shape inference
  - Add testcases which include dynamic shape
  - Support shape inference with unknown dimensions
  - Reduce code duplication by grouping  similar logic as a helper function
- Type inference
  - Reduce code duplication by grouping similar logic as a helper function

## How

1. Select an operation/operations to implement and write your name in the list of table at **assignee**.
    - You can select multiple operations which logic is similar so that PR can be reviewed at once. (ex: ReLU and ReLU6)
    - **Only reserve as much as you can finish in one week. DO NOT BE GREEDY.**
1. Implement each operations as following rules
    - Shape Inference
        - Write new recipes with dynamic shape, which are enough to cover most of algorithm codes. (For code coverage)
        - Add the recipes to `common-artifacts/exclude.lst`. (TestDataGenerator not supports yet)
        - Add the recipes to `res/TensorFlowLiteRecipes/` or `res/CircleRecipes`.
        - Upgrade shape inference algorithm with considering unknown dimension. (Group similar algorithms as helper function)
    - Type Inference
        - Move original type inference algorithm. (Group similar algorithms as helper function)
1. Create PR(s). Then write the PR number in the table at
    - **PR(exclude)**, for adding recipe to `common-artifacts/exclude.lst`. If not needed, write :no_entry_sign: ans skip it.
    - **PR(res)**, for adding recipe to `res/`. If not needed, write :no_entry_sign: ans skip it.
    - **PR(rule)**, for inference algorithm implementation
1. After PR is merged, write :heavy_check_mark: at **Status** in the table.

### Examples (Guide)

TBD

### Background

This issue is originated from #5501. Please refer to the [comment in that issue](https://github.com/Samsung/ONE/issues/5501#issue-775808564).

**Comment by llFreetimell** [comment]
### Operation Table (A-C)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(exlucde) | PR(res) | PR(algorithm) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| Abs |      |      |      |      |      |
| Add |      |      |      |      |      |
| AddN |      |      |      |      |      |
| ArgMax |      |      |      |      |      |
| ArgMin |      |      |      |      |      |
| AveragePool2D |      |      |      |      |      |
| BatchMatMul |      |      |      |      |      |
| BatchToSpaceND |      |      |      |      |      |
| Cast |      |      |      |      |      |
| Ceil |      |      |      |      |      |
| Concatenation |      |      |      |      |      |
| Const |      |      |      |      |      |
| Conv2D |      |      |      |      |      |
| Cos |      |      |      |      |      |
| Custom |      |      |      |      |      |

**Comment by llFreetimell** [comment]
### Operation Table (D-F)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(exlucde) | PR(res) | PR(algorithm) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| DepthToSpace |      |      |      |      |      |
| DepthwiseConv2D |      |      |      |      |      |
| Dequantize |      |      |      |      |      |
| Div |      |      |      |      |      |
| Elu |      |      |      |      |      |
| Equal |      |      |      |      |      |
| Exp |      |      |      |      |      |
| ExpandDims |      |      |      |      |      |
| FakeQuant |      |      |      |      |      |
| Fill |      |      |      |      |      |
| Floor |      |      |      |      |      |
| FloorDiv |      |      |      |      |      |
| FloorMod |      |      |      |      |      |
| FullyConnected |      |      |      |      |      |

**Comment by llFreetimell** [comment]
### Operation Table (G-L)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(exlucde) | PR(res) | PR(algorithm) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| Greater |      |      |      |      |      |
| GreaterEqual |      |      |      |      |      |
| If |      |      |      |      |      |
| L2Normalize |      |      |      |      |      |
| L2Pool2D |      |      |      |      |      |
| LeakyRelu |      |      |      |      |      |
| Less |      |      |      |      |      |
| LessEqual |      |      |      |      |      |
| LocalResponseNormalization |      |      |      |      |      |
| Log |      |      |      |      |      |
| LogicalAnd |      |      |      |      |      |
| LogicalNot |      |      |      |      |      |
| LogicalOr |      |      |      |      |      |
| Logistic |      |      |      |      |      |
| LogSoftmax |      |      |      |      |      |

**Comment by llFreetimell** [comment]
### Operation Table (M-P)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(exlucde) | PR(res) | PR(algorithm) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| MatrixDiag |      |      |      |      |      |
| MatrixSetDiag |      |      |      |      |      |
| Maximum |      |      |      |      |      |
| MaxPool2D |      |      |      |      |      |
| Mean |      |      |      |      |      |
| Minimum |      |      |      |      |      |
| MirrorPad |      |      |      |      |      |
| Mul |      |      |      |      |      |
| Neg |      |      |      |      |      |
| NonMaxSuppressionV4 |      |      |      |      |      |
| NonMaxSuppressionV5 |      |      |      |      |      |
| NotEqual |      |      |      |      |      |
| OneHot |      |      |      |      |      |
| Pack |      |      |      |      |      |
| Pad |      |      |      |      |      |
| PadV2 |      |      |      |      |      |
| Pow |      |      |      |      |      |
| PRelu |      |      |      |      |      |

**Comment by llFreetimell** [comment]
### Operation Table (R)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(exlucde) | PR(res) | PR(algorithm) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| Range |      |      |      |      |      |
| Rank |      |      |      |      |      |
| ReduceAny |      |      |      |      |      |
| ReduceMax |      |      |      |      |      |
| ReduceMin |      |      |      |      |      |
| ReduceProd |      |      |      |      |      |
| Relu |      |      |      |      |      |
| Relu6 |      |      |      |      |      |
| ReluN1To1 |      |      |      |      |      |
| Reshape |      |      |      |      |      |
| ResizeBilinear |      |      |      |      |      |
| ResizeNearestNeighbor |      |      |      |      |      |
| ReverseSequence |      |      |      |      |      |
| ReverseV2 |      |      |      |      |      |
| Round |      |      |      |      |      |
| Rsqrt |      |      |      |      |      |

**Comment by llFreetimell** [comment]
### Operation Table (S)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(exlucde) | PR(res) | PR(algorithm) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| ScatterNd |      |      |      |      |      |
| SegmentSum |      |      |      |      |      |
| Select |      |      |      |      |      |
| SelectV2 |      |      |      |      |      |
| Shape |      |      |      |      |      |
| Sin |      |      |      |      |      |
| Slice |      |      |      |      |      |
| Softmax |      |      |      |      |      |
| SpaceToBatchND |      |      |      |      |      |
| SpaceToDepth |      |      |      |      |      |
| SparseToDense |      |      |      |      |      |
| Split |      |      |      |      |      |
| SplitV |      |      |      |      |      |
| Sqrt |      |      |      |      |      |
| Square |      |      |      |      |      |
| SquaredDifference |      |      |      |      |      |
| Squeeze |      |      |      |      |      |
| StridedSlice |      |      |      |      |      |
| Sub |      |      |      |      |      |
| Sum |      |      |      |      |      |

**Comment by llFreetimell** [comment]
### Operation Table (T-Z)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(exlucde) | PR(res) | PR(algorithm) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| Tanh |      |      |      |      |      |
| Tile |      |      |      |      |      |
| TopKV2 |      |      |      |      |      |
| Transpose |      |      |      |      |      |
| TransposeConv |      |      |      |      |      |
| UnidirectionalSequenceLSTM |      |      |      |      |      |
| Unique |      |      |      |      |      |
| Unpack |      |      |      |      |      |
| Where |      |      |      |      |      |
| While |      |      |      |      |      |
| ZerosLike |      |      |      |      |      |

**Comment by llFreetimell** [comment]
### Operation Table (Circle Only Ops)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(exlucde) | PR(res) | PR(algorithm) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| BCQFullyConnected |      |      |      |      |      |
| BCQGather |      |      |      |      |      |
| InstanceNorm |      |      |      |      |      |

**Comment by llFreetimell** [comment]
### Operation Table (Virtual Ops)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(exlucde) | PR(res) | PR(algorithm) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| Input |      | :no_entry_sign: | :no_entry_sign: |  |      |
| Output |      |:no_entry_sign: | :no_entry_sign: |  |      |
| OutputDummy |    |:no_entry_sign:| :no_entry_sign: |  |      |
| OutputExclude |       |:no_entry_sign: | :no_entry_sign: |  |      |
| CustomOut |      | :no_entry_sign:  | :no_entry_sign: |  |      |
| IfOut |      | :no_entry_sign:  | :no_entry_sign: |  |      |
| NonMaxSuppressionV4Out |      | :no_entry_sign:  | :no_entry_sign: |  |      |
| NonMaxSuppressionV5Out |      | :no_entry_sign:  | :no_entry_sign: |  |      |
| SplitOut |      | :no_entry_sign:  | :no_entry_sign: |  |      |
| SplitVOut |      | :no_entry_sign:  | :no_entry_sign: |  |      |
| TopKV2Out |      | :no_entry_sign:  | :no_entry_sign: |  |      |
| UniqueOut |      | :no_entry_sign:  | :no_entry_sign: |  |      |
| UnpackOut |      | :no_entry_sign:  | :no_entry_sign: |  |      |
| WhileOut |      | :no_entry_sign:  | :no_entry_sign: |  |      |
